spipe/main.c

```diff
@@ -17,15 +17,25 @@
 #include "pushbits.h"
 
 static int
-callback_conndied(void * cookie)
+callback_conndied(void * cookie, int reason)
 {
 	int * conndone = cookie;
 
+	/* Check reason. */
+	if (reason == PROTO_CONN_CONNECT_FAILED) {
+		warn0("Could not connect");
+		goto err0;
+	}
+
 	/* Quit event loop. */
 	*conndone = 1;
 
 	/* Success! */
 	return (0);
+
+err0:
+	/* Failed */
```

Comment:
I wouldn't do this.  Normally when a callback returns `-1` it means "something drastic and unrecoverable happened", e.g., not being able to allocate memory for a continuation or a syscall failing which normally shouldn't.  Networks not working is not a *catastrophic* failure.

So as far as making `main` return 1, I'd go with "record that the connection could not be established and decide what value to exit from main with" rather than "throw a callback failure and watch everything abort".

As far as the UX goes, this means seeing "Could not connect" and ending with `exit(1)` but not reporting "Error running event loop".